inbox/auth/utils.py

```diff
@@ -37,42 +37,31 @@ def auth_requires_app_password(exc: IMAPClient.Error) -> bool:
     )
 
 
-# IMAP doesn't really have error semantics, so we have to match the error
-# message against a list of known response strings to determine whether we
-# couldn't log in because the credentials are invalid, or because of some
-# temporary server error.
-AUTH_INVALID_PREFIXES = [
-    prefix.lower()
-    for prefix in (
-        "[authenticationfailed]",
-        "incorrect username or password",
-        "invalid login or password",
-        "login login error password error",
-        "[auth] authentication failed.",
-        "invalid login credentials",
-        "[ALERT] Please log in via your web browser",
-        "LOGIN Authentication failed",
-        "authentication failed",
-        "[ALERT] Invalid credentials(Failure)",
-        "Invalid email login",
-        "failed: Re-Authentication Failure",
-        "Invalid",
-        "Login incorrect",
-        "LOGIN GroupWise login failed",
-        "authentication failed",
-        "LOGIN bad",  # LOGIN bad username or password
-        "[AUTHORIZATIONFAILED]",
-        "incorrect password",
+# IMAP doesn't have error semantics, so we have to match the error message
+# against a list of known responses to determine whether we couldn't log in
+# because the credentials are invalid. Sometimes a single pattern matches many
+# different responses from servers. For real error messages that have been
+# received from servers in the past, see the test suite.
+AUTH_INVALID_PATTERNS = [
+    pattern.lower()
+    for pattern in (
+        "fail",
```

Comment:
Isn't this function about determining if the error message says that authentication failed due to invalid credentials?

I imagine an error message could contain "fail" or "bad" for other reasons?

Reply:
I read the entire PR description before reviewing. It's not about appeasing me, it's about making sure this logic doesn't return false positives. I can imagine an error like "Temporary server failure" that would be classified as an authentication error, while it shouldn't be.

Reply:
That makes sense. I added some qualifiers for `fail`, and I also added a qualifier for `bad` since it could match the common "Bad Gateway" error. I don't think "invalid" or "incorrect" need qualifiers, but let me know if you disagree.

Done in https://github.com/closeio/sync-engine/pull/1043/commits/f8d9b3d97eaa11d85f6c29cd72a5b08eb9057ae2.